### PR TITLE
Signoff integration and improvements

### PR DESCRIPTION
**Pull request by harindup**
**DONE**

- Searching for a group to sign off
- Signing off (double click on a cell or assignment header)
- Comment prompts as necessary for changes/insufficient sign offs

**TODO**

- Styling (looks very meh)
- Comments integration
- Sign off history viewing

### CODE REVIEW

File: frontend/src/styling/index.scss [hunk 1]
```diff
@@ -207,20 +208,76 @@ button {
 }
 
 .sign-off-table {
+    order: 0.25px solid rgba(0, 0, 0, 0.25);
```
Comment (JustinPraas) [inline frontend/src/styling/index.scss]: Wait, how does this work and what actually does it do?

Besides, when I looked it up, it seems there might be problems for Safari users. (https://www.w3schools.com/cssref/css3_pr_order.asp)

Add this:

> -webkit-order: 0.25px solid rgba(0, 0, 0, 0.25);

The above only applies if we are talking about the same `order` thing..

File: frontend/src/state/sign-off/sagas.ts [hunk 2]
```diff
@@ -1,53 +1,57 @@
...
-            `assignmentSets/${action.asid}`);
+        const [signoffs, group, assignmentSet ] = yield all(
+            [
+                call(authenticatedFetchJSON, "GET",
+                    `courses/${action.cid}/signoffresults`, { groupId: action.gid, assignmentSetId: action.asid }),
+                call(authenticatedFetchJSON, "GET", `groups/${action.gid}`),
+                call(authenticatedFetchJSON, "GET", `assignmentSets/${action.asid}`),
+            ],
+        );
+
```
[the remark this answers is not in the document]
Comment (JustinPraas) [reply frontend/src/state/sign-off/sagas.ts]: Refering to the yield all(...) things, which the above snippet doesn't show

File: frontend/src/components/pages/sign-off/SignoffTable.tsx [hunk 3]
```diff
@@ -4,180 +4,508 @@ import {
...
+    signoffs: SignOffResultDtoCompact[] | null;
     group: GroupDtoFull | null;
     assignmentSet: AssignmentSetDtoFull | null;
     localChanges: SignOffChange[] | null;
     saving: boolean;
 
-    requestSignOffs: (asid: number, cid: number, gid: number) => {
+    fetchSignOffs: (asid: number, cid: number, gid: number) => {
         type: string,
     };
```
[the remark this answers is not in the document]
Comment (harindup) [reply frontend/src/components/pages/sign-off/SignoffTable.tsx]: We should indeed. That part isn't properly typed ATM. I might just fix these stuff for signoff in this PR itself.

File: frontend/src/components/pages/sign-off/SignoffTable.tsx [hunk 4]
```diff
@@ -4,180 +4,508 @@ import {
...
         }
+    }
 
-        switch (remoteSignoff.result) {
-            case "COMPLETE":
-                return SignOff.Complete;
+    /**
+     * Commit pending changes to state and pushed to backend afterwards.
+     * @param changes sign off changes to commit to state and push to backend
+     */
```
[the remark this answers is not in the document]
Comment (harindup) [reply frontend/src/components/pages/sign-off/SignoffTable.tsx]: I'll do that. I agree that it's nice to have *some* convention at least. Google isn't authoritative, but at least it's quite extensive.

File: frontend/src/state/sign-off/sagas.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (harindup) [reply frontend/src/state/sign-off/sagas.ts]: Parallel API calls. Yield is "blocking". I need all three datasets, but there's no need to wait for one to fetch the other. This speeds up things a bit.

File: frontend/src/styling/index.scss (hunk 1 above)
Comment (harindup) [reply frontend/src/styling/index.scss]: That's a typo; sorry. It should be "border".

Comment (JustinPraas) [reply frontend/src/styling/index.scss]: Oh now.. that explains a lot

File: frontend/prettier.config.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+    singleQuote: false,
+    trailingComma: 'all',
+    arrowParens: 'always'
+  };
```
Comment (JustinPraas) [inline frontend/prettier.config.js]: Is it possible to have indentation of 4 spaces instead of 2?

Comment (harindup) [reply frontend/prettier.config.js]: That's indeed possible. I'll change.

File: frontend/src/components/pages/sign-off/SignOffSearch.tsx [hunk 6]
```diff
@@ -68,100 +82,152 @@ class SignOffSearch extends Component<SignOffSearchProps & RouteComponentProps<a
...
+            Number(this.props.match.params.asid),
+        );
         return (
-            <Dropdown isOpen={this.state.dropdownOpen}
-                toggle={this.toggle}
-                className="p-4">
+            <Dropdown isOpen={this.state.dropdownOpen} toggle={this.toggle}>
                 <DropdownToggle caret>
-                    {selectedAssignmentSet !== null ? selectedAssignmentSet.name : "Select an assignment set"}
+                    {selectedAssignmentSet !== null
```
Comment (JustinPraas) [inline frontend/src/components/pages/sign-off/SignOffSearch.tsx]: !== null

File: frontend/src/components/pages/sign-off/SignOffSearch.tsx [hunk 7]
```diff
@@ -68,100 +82,152 @@ class SignOffSearch extends Component<SignOffSearchProps & RouteComponentProps<a
...
         );
     }
 
     private renderSearchBar() {
         return (
             <Autosuggest
-                suggestions={this.props.searchResult !== null
-                    ? this.orderSearchResults(this.props.searchResult) : []}
+                suggestions={
+                    this.props.searchResult !== null
```
Comment (JustinPraas) [inline frontend/src/components/pages/sign-off/SignOffSearch.tsx]: !== null

File: frontend/src/components/pages/sign-off/SignOffSearch.tsx [hunk 8]
```diff
@@ -68,100 +82,152 @@ class SignOffSearch extends Component<SignOffSearchProps & RouteComponentProps<a
...
             />
         );
     }
 
-    private orderSearchResults(searchResult: GroupAssignmentSetCombination[]): GroupAssignmentSetCombination[] {
+    private orderSearchResults(
+        searchResult: GroupAssignmentSetCombination[],
+    ): GroupAssignmentSetCombination[] {
         const asid = Number(this.props.match.params.asid);
         if (asid !== undefined) {
```
Comment (JustinPraas) [inline frontend/src/components/pages/sign-off/SignOffSearch.tsx]: The thing with !isNaN and > 0

File: frontend/src/components/pages/sign-off/SignOffSearch.tsx [hunk 9]
```diff
@@ -172,7 +238,7 @@ class SignOffSearch extends Component<SignOffSearchProps & RouteComponentProps<a
         }
     }
 
-    private getAssignmentSetByID(asid: number): (AssignmentSetDtoBrief | null) {
+    private getAssignmentSetByID(asid: number): AssignmentSetDtoBrief | null {
         if (asid !== null && this.props.assignmentSets() !== null) {
```
Comment (JustinPraas) [inline frontend/src/components/pages/sign-off/SignOffSearch.tsx]: The thing with !isNaN and > 0

File: frontend/src/components/pages/sign-off/SignoffTable.tsx [hunk 10]
```diff
@@ -1,183 +1,589 @@
...
+    private hasChangedInSession(pid: number, aid: number): boolean {
+        return (
+            this.state.changes.findIndex(
+                (c) => c.aid === aid && c.pid === pid,
+            ) >= 0
+        );
+    }
+
+    /**
+     * Reload component data.
```
Comment (JustinPraas) [inline frontend/src/components/pages/sign-off/SignoffTable.tsx]: Reload_s_

**Pull request merged**